### PR TITLE
Undo redo trace

### PR DESCRIPTION
**Pull request by potatowagon**
print undoables and redoables in command box. Tested with UI

### CODE REVIEW

**Comment by pyokagan** [comment]
@potatowagon What's the purpose of this PR?

**Comment by pyokagan** [comment]
@potatowagon What's the status of this PR?

**Comment by potatowagon** [comment]
Test error for undo is: HeadAtBoundary Exception not thrown, but i cant figure out why... UI shows that it is thrown by displaying No actions to Undo in command box

**Comment by potatowagon** [comment]
Redo passes the throw HeadAtBoundaryException test

File: src/main/java/seedu/address/model/Model.java [hunk 1]
```diff
@@ -124,4 +126,35 @@
...
+     */
+    Command undo() throws HeadAtBoundaryException;
+
+    /**
+     *@return true if TaskBook has changed
+     *Unlikely the commits list will be empty since recordState() is called when ModelManager is initialised. Thus there will always be at least one Commit in commits.
+     */
+    boolean hasUncommittedChanges();
+
+    String printRedoables();
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/Model.java:156]: I hope this is just a debug method that will be removed when this PR is done.

File: src/test/java/seedu/address/model/ModelTest.java [hunk 2]
```diff
@@ -1,20 +1,28 @@
...
 import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.logic.commands.ClearCommand;
+import seedu.address.logic.commands.Command;
+import seedu.address.logic.commands.CommandResult;
+import seedu.address.model.ModelManager.HeadAtBoundaryException;
 import seedu.address.model.task.TypicalDeadlineTasks;
 import seedu.address.model.task.TypicalEventTasks;
 import seedu.address.model.task.TypicalFloatingTasks;
 
+@SuppressWarnings("deprecation")
```
Comment (pyokagan) [inline src/test/java/seedu/address/model/ModelTest.java:25]: Maybe replace all the `Assert.fail()` with `assert false : "Should not be called"`

Comment (pyokagan) [reply src/test/java/seedu/address/model/ModelTest.java:25]: And then remove this `@SuppressWarnings`

File: src/main/java/seedu/address/model/ModelManager.java [hunk 3]
```diff
@@ -265,4 +276,89 @@ public void setDeadlineTaskFilter(Predicate<? super DeadlineTask> predicate) {
...
+        String trace = "";
+        for (int i = head; i < commits.size() ; i++) {
+            trace += commits.get(i).getCommand() + "\n";
+        }
+        return trace;
+    }
+
+    private class Commit {
+        private TaskBook taskBook;
+        private Command command;
```
Comment (pyokagan) [inline src/main/java/seedu/address/model/ModelManager.java:343]: Looks like these fields are immutable and can be declared `final`

File: src/main/java/seedu/address/model/Model.java (hunk 1 above)
Comment (pyokagan) [reply src/main/java/seedu/address/model/Model.java:156]: Hmmm, maybe leave the undo trace and redo trace commands these rounds. If you want to implement them the Model interface will need to be rethought. (We will need to find some way to expose the `Commit` objects cleanly).

**Comment by pyokagan** [comment]
@potatowagon Rebase this branch asap.

**Comment by potatowagon** [comment]
okay

**Comment by potatowagon** [comment]
could you help me do the rebase on your machine? I dont know which merge conflicts to keep. Also i tried rebasing but cannot

screenshot 51

**Comment by pyokagan** [comment]
Yeah, your branch is too far behind so there's too many conflicts. I just squashed the changes and redid the commit: #134 

Next time rebase more regularly.

**Comment by potatowagon** [comment]
thanks! im halfway rebasing alrd, will still pr when done

**Comment by potatowagon** [comment]
oh great its done!

**Comment by potatowagon** [comment]
extremely weird, git stopped at rebase 12/21 and says success

**Comment by potatowagon** [comment]
oh wait yeah its done :)

**Pull request closed without merging**